In [1]:
import numpy as np
import astroquery
from astropy import units as u
from astroquery.simbad import Simbad
from astropy.coordinates import SkyCoord
import pandas as pd
from astropy.table import Table, vstack

In [19]:
radius=2.
df_2 = pd.read_csv('/home/nicolas/nico/Data/Masivas/Data_OB_stars/OBAGDR3.csv')


In [20]:
df_2 = df_2.head(100)

In [28]:
Simbad.TIMEOUT = 100000
Simbad.ROW_LIMIT = 0

In [29]:
Simbad.add_votable_fields('sptype')
Simbad.add_votable_fields('sp_bibcode')
coordinates = SkyCoord(df_2[["ra","dec"]].values,unit=("deg","deg"))
tabF = Simbad.query_region(coordinates, radius=radius*u.arcsec)

In [30]:
table=tabF.to_pandas()

In [31]:
table

,MAIN_ID,RA,DEC,RA_PREC,DEC_PREC,COO_ERR_MAJA,COO_ERR_MINA,COO_ERR_ANGLE,COO_QUAL,COO_WAVELENGTH,...,SP_BIBCODE_8,SP_TYPE_5,SP_QUAL_5,SP_BIBCODE_9,SP_BIBCODE_10,SP_TYPE_6,SP_QUAL_6,SP_BIBCODE_11,SP_BIBCODE_12,SCRIPT_NUMBER_ID
0,Gaia DR2 5188182749674777984,11 48 31.0247,-89 32 07.358,14,14,0.0267,0.0258,90,A,O,...,,,,,,,,,,2
1,Gaia DR3 6341129165091554304,21 28 26.1577,-89 09 42.951,14,14,0.0162,0.0161,90,A,O,...,,,,,,,,,,6
2,Gaia DR2 5188434091160619136,06 01 52.8146,-89 05 15.739,14,14,0.0185,0.0201,90,A,O,...,,,,,,,,,,9
3,* sig Oct,21 08 46.8635,-88 57 23.398,14,14,0.0516,0.0605,90,A,O,...,1975MSS...C01....0H,F0IV,C,1975MSS...C01....0H,1975MSS...C01....0H,F0IV,C,1975MSS...C01....0H,1975MSS...C01....0H,10
4,HD 108971,12 42 10.6916,-88 33 29.040,14,14,0.0435,0.0388,90,A,O,...,1975MSS...C01....0H,G0,E,1975MSS...C01....0H,1975MSS...C01....0H,G0,E,1975MSS...C01....0H,1975MSS...C01....0H,17
5,V* RW Oct,18 29 04.7062,-88 32 32.143,14,14,0.0108,0.0121,90,A,O,...,,,,,,,,,,18
6,TYC 9521-341-1,15 38 32.9741,-88 26 15.573,14,14,0.5431,0.5786,90,A,O,...,,,,,,,,,,22
7,Gaia DR2 5188570464962270336,08 53 13.3274,-88 26 09.852,14,14,0.0130,0.0132,90,A,O,...,,,,,,,,,,23
8,HD 1348,00 12 33.9573,-88 21 46.334,14,14,0.0196,0.0180,90,A,O,...,1975MSS...C01....0H,B9.5IV,C,1975MSS...C01....0H,1975MSS...C01....0H,B9.5IV,C,1975MSS...C01....0H,1975MSS...C01....0H,24
9,HD 16477,01 49 11.7717,-88 21 26.725,14,14,0.0339,0.0345,90,A,O,...,1975MSS...C01....0H,B9/9.5V,D,1975MSS...C01....0H,1975MSS...C01....0H,B9/9.5V,D,1975MSS...C01....0H,1975MSS...C01....0H,26


In [63]:
table[table.duplicated(subset='SCRIPT_NUMBER_ID')==True]['SCRIPT_NUMBER_ID'].values

array([  22,   58,  106,  286,  323,  323,  426,  479,  479,  526,  597,
        597,  621,  621,  660,  805,  805,  854,  854,  921, 1094, 1165,
       1225, 1225, 1239, 1258, 1275, 1281, 1281, 1373, 1373, 1374, 1412,
       1804, 1804, 1852, 1852, 1868, 1868, 1942, 1942, 1975, 2125, 2173,
       2273, 2273, 2288, 2288, 2289, 2309, 2312, 2340, 2340, 2391, 2391,
       2431, 2578, 2578, 2633, 2651, 2661, 2670, 2683, 2683, 2728, 2728,
       2765, 2782, 2782, 2794, 2794, 2805, 2805, 2849, 2849, 2885, 2885,
       2948, 2948, 2957, 3020, 3020, 3023, 3023, 3066, 3066, 3144, 3146,
       3203, 3203, 3246, 3246, 3255, 3255, 3399, 3498, 3648, 3728, 3883,
       3923, 4013, 4072, 4101, 4101, 4164, 4164, 4181, 4181, 4219, 4219,
       4230, 4252, 4283, 4291, 4291, 4311, 4320, 4320, 4378, 4416, 4416,
       4450, 4530, 4578, 4602, 4623, 4623, 4649, 4800, 4914, 4914, 4961,
       5095, 5116, 5116, 5210, 5273, 5338, 5356, 5356, 5401, 5469, 5496,
       5521, 5601, 5687, 5746, 5753, 5762, 5820, 58

In [68]:
table[table['SCRIPT_NUMBER_ID']==22][['SP_TYPE','SP_BIBCODE']]

,SP_TYPE,SP_BIBCODE
21,F8,1993yCat.3135....0C
22,,


In [69]:
table[table['SCRIPT_NUMBER_ID']==58][['SP_TYPE','SP_BIBCODE']]

,SP_TYPE,SP_BIBCODE
58,G0,1993yCat.3135....0C
59,,


In [70]:
table[table['SCRIPT_NUMBER_ID']==106][['SP_TYPE','SP_BIBCODE']]

,SP_TYPE,SP_BIBCODE
107,A0,1993yCat.3135....0C
108,,


In [71]:
for i in table[table.duplicated(subset='SCRIPT_NUMBER_ID')==True]['SCRIPT_NUMBER_ID'].values:
    print(table[table['SCRIPT_NUMBER_ID']==i][['SP_TYPE','SP_BIBCODE']])

   SP_TYPE           SP_BIBCODE
21      F8  1993yCat.3135....0C
22                             
   SP_TYPE           SP_BIBCODE
58      G0  1993yCat.3135....0C
59                             
    SP_TYPE           SP_BIBCODE
107      A0  1993yCat.3135....0C
108                             
    SP_TYPE           SP_BIBCODE
288      A0  1993yCat.3135....0C
289      A0                     
    SP_TYPE           SP_BIBCODE
326      A0  1993yCat.3135....0C
327                             
328                             
    SP_TYPE           SP_BIBCODE
326      A0  1993yCat.3135....0C
327                             
328                             
    SP_TYPE           SP_BIBCODE
431      B8                     
432      B8  1993yCat.3135....0C
    SP_TYPE           SP_BIBCODE
485      K0  1995A&AS..110..367N
486                             
487                             
    SP_TYPE           SP_BIBCODE
485      K0  1995A&AS..110..367N
486                             
487             

In [75]:
obj_rep=table[table.duplicated(subset='SCRIPT_NUMBER_ID')==True]['SCRIPT_NUMBER_ID'].values
len(np.unique(obj_rep))

110

We queried Simbad with 2arsec, and inspected objects with two counterparts or more in this radius (110 objects). For all cases at least one of the counterparts did not have a reference for the spectral type or spectral type at all and therefore we discarded them from any further analysis

In [79]:
tableNR=table.drop_duplicates(subset='SCRIPT_NUMBER_ID',keep=False)

In [81]:
tableNR.shape

(5946, 39)

In [83]:
np.sum(tableNR['SP_BIBCODE'].values=='')

904

In [90]:
tableNR.groupby(['SP_BIBCODE']).count()['MAIN_ID'].sort_values()

SP_BIBCODE
2010A&A...515A..16S       1
2008ApJ...672..558C       1
2009A&A...495..121M       1
2009ApJ...695...75H       1
2010A&A...516A..35H       1
                       ... 
2008AJ....135..564B     264
2020ApJS..249....3A     274
                        904
1995A&AS..110..367N    1490
1993yCat.3135....0C    2182
Name: MAIN_ID, Length: 92, dtype: int64

In [91]:
(904+1490+2182)/5946

0.7695930036999664

In [26]:
test=df_2[0:6000]

In [27]:
Simbad.add_votable_fields('sptype')
Simbad.add_votable_fields('sp_bibcode')
coordinates = SkyCoord(test[["ra","dec"]].values,unit=("deg","deg"))
tab = Simbad.query_region(coordinates, radius=radius*u.arcsec)

In [28]:
test2=df_2[6000:]
coordinates = SkyCoord(test2[["ra","dec"]].values,unit=("deg","deg"))
tab2 = Simbad.query_region(coordinates, radius=radius*u.arcsec)

In [30]:
tabla=vstack([tab, tab2])

In [31]:
tabla

MAIN_ID,RA,DEC,RA_PREC,DEC_PREC,COO_ERR_MAJA,COO_ERR_MINA,COO_ERR_ANGLE,COO_QUAL,COO_WAVELENGTH,COO_BIBCODE,SP_TYPE,SP_QUAL,SP_BIBCODE,SP_TYPE_2,SP_QUAL_2,SP_BIBCODE_2,SP_BIBCODE_3,SP_TYPE_3,SP_QUAL_3,SP_BIBCODE_4,SP_BIBCODE_5,SP_TYPE_4,SP_QUAL_4,SP_BIBCODE_6,SP_BIBCODE_7,SP_TYPE_5,SP_QUAL_5,SP_BIBCODE_8,SP_BIBCODE_9,SP_TYPE_6,SP_QUAL_6,SP_BIBCODE_10,SP_BIBCODE_11,SCRIPT_NUMBER_ID
,"""h:m:s""","""d:m:s""",,,mas,mas,deg,,,,,,,,,,,,,,,,,,,,,,,,,,,
object,str13,str13,int16,int16,float32,float32,int16,str1,str1,object,object,str1,object,object,str1,object,object,object,str1,object,object,object,str1,object,object,object,str1,object,object,object,str1,object,object,int32
EC 21086-8739,21 28 07.0677,-87 27 05.467,14,14,0.014,0.016,90,A,O,2020yCat.1350....0G,sdB,D,2013MNRAS.431..240O,sdB,D,2013MNRAS.431..240O,2013MNRAS.431..240O,sdB,D,2013MNRAS.431..240O,2013MNRAS.431..240O,sdB,D,2013MNRAS.431..240O,2013MNRAS.431..240O,sdB,D,2013MNRAS.431..240O,2013MNRAS.431..240O,sdB,D,2013MNRAS.431..240O,2013MNRAS.431..240O,1
EC 03379-8354,03 32 14.6120,-83 44 13.060,14,14,0.016,0.017,90,A,O,2020yCat.1350....0G,sdB+F,D,2013MNRAS.431..240O,sdB+F,D,2013MNRAS.431..240O,2013MNRAS.431..240O,sdB+F,D,2013MNRAS.431..240O,2013MNRAS.431..240O,sdB+F,D,2013MNRAS.431..240O,2013MNRAS.431..240O,sdB+F,D,2013MNRAS.431..240O,2013MNRAS.431..240O,sdB+F,D,2013MNRAS.431..240O,2013MNRAS.431..240O,2
CPD-82 801,20 20 20.5286,-82 25 39.586,14,14,0.010,0.011,90,A,O,2020yCat.1350....0G,A,E,2013MNRAS.431..240O,A,E,2013MNRAS.431..240O,2013MNRAS.431..240O,A,E,2013MNRAS.431..240O,2013MNRAS.431..240O,A,E,2013MNRAS.431..240O,2013MNRAS.431..240O,A,E,2013MNRAS.431..240O,2013MNRAS.431..240O,A,E,2013MNRAS.431..240O,2013MNRAS.431..240O,3
EC 23073-7751,23 11 01.3018,-77 35 01.144,14,14,0.018,0.021,90,A,O,2020yCat.1350....0G,A,E,2013MNRAS.431..240O,A,E,2013MNRAS.431..240O,2013MNRAS.431..240O,A,E,2013MNRAS.431..240O,2013MNRAS.431..240O,A,E,2013MNRAS.431..240O,2013MNRAS.431..240O,A,E,2013MNRAS.431..240O,2013MNRAS.431..240O,A,E,2013MNRAS.431..240O,2013MNRAS.431..240O,4
V* AE Oct,20 22 55.2378,-76 04 04.806,14,14,0.010,0.011,90,A,O,2020yCat.1350....0G,F5,D,2013MNRAS.431..240O,F5,D,2013MNRAS.431..240O,2013MNRAS.431..240O,F5,D,2013MNRAS.431..240O,2013MNRAS.431..240O,F5,D,2013MNRAS.431..240O,2013MNRAS.431..240O,F5,D,2013MNRAS.431..240O,2013MNRAS.431..240O,F5,D,2013MNRAS.431..240O,2013MNRAS.431..240O,5
EC 04002-7554,03 58 57.3037,-75 46 33.084,14,14,0.025,0.022,90,A,O,2020yCat.1350....0G,B3,D,2013MNRAS.431..240O,B3,D,2013MNRAS.431..240O,2013MNRAS.431..240O,B3,D,2013MNRAS.431..240O,2013MNRAS.431..240O,B3,D,2013MNRAS.431..240O,2013MNRAS.431..240O,B3,D,2013MNRAS.431..240O,2013MNRAS.431..240O,B3,D,2013MNRAS.431..240O,2013MNRAS.431..240O,6
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
V* V462 Cep,23 51 49.6549,+81 50 12.659,14,14,0.016,0.014,90,A,O,2020yCat.1350....0G,A2,E,1993yCat.3135....0C,A2,E,1993yCat.3135....0C,1993yCat.3135....0C,A2,E,1993yCat.3135....0C,1993yCat.3135....0C,A2,E,1993yCat.3135....0C,1993yCat.3135....0C,A2,E,1993yCat.3135....0C,1993yCat.3135....0C,A2,E,1993yCat.3135....0C,1993yCat.3135....0C,6051


In [35]:
tabla['SCRIPT_NUMBER_ID']

1
2
3
4
5
6
7
8
9
...
6049


In [34]:
Simbad.query_region?

In [21]:
np.sum(tab['SP_BIBCODE'].value=='1974A&AS...16..277C')

11

In [22]:
np.sum(tab['SP_BIBCODE'].value=='')

14

In [11]:
test

,source_id,ra,dec,ra_error,dec_error,parallax,parallax_error,parallax_over_error,pmra,pmdec,...,R_x,J,H,K,u,g,r_xa,i,z,angDist
0,6341936069186773888,322.029689,-87.451542,0.014296,0.016214,0.061743,0.018665,3.307911,2.399615,-5.225902,...,14.489,14.187,14.238,14.337,NaN,14.242,14.337,NaN,NaN,0.092008
1,4615556475452740480,53.061042,-83.736951,0.016101,0.016573,0.046835,0.017140,2.732576,3.895391,2.275492,...,14.375,14.340,14.332,14.381,NaN,14.058,14.251,14.482,NaN,0.072181
2,6348968408118750208,305.085692,-82.427702,0.009958,0.010758,0.837793,0.012066,69.435165,4.614924,-8.705492,...,NaN,10.080,9.876,9.788,NaN,NaN,NaN,NaN,NaN,0.157649
3,6354071482101259648,347.755530,-77.583660,0.018343,0.020622,0.117265,0.021830,5.371705,5.113016,-2.043968,...,NaN,14.383,14.336,14.120,NaN,NaN,NaN,NaN,NaN,0.088103
4,6363504054917758592,305.730093,-76.068019,0.009825,0.011417,0.232584,0.013630,17.064213,-3.499222,-3.919243,...,NaN,12.770,12.571,12.506,NaN,NaN,NaN,NaN,NaN,0.084065
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5995,3132997570911033984,103.681461,7.027954,0.017873,0.017061,1.750222,0.021325,82.074230,-1.634334,-2.223650,...,NaN,8.776,8.765,8.724,NaN,NaN,NaN,NaN,NaN,0.044154
5996,4297132032545725056,299.629678,7.031396,0.362252,0.301926,0.773246,0.535641,1.443591,-0.622325,-5.720377,...,NaN,10.065,10.135,10.026,NaN,NaN,NaN,NaN,NaN,0.092066
5997,3153780741558513280,105.596462,7.035489,0.030995,0.027926,1.333713,0.033379,39.956380,-1.599291,-1.653234,...,NaN,8.667,8.736,8.708,NaN,NaN,NaN,NaN,NaN,0.036803
5998,3322161296563994752,92.282647,7.055391,0.027482,0.032031,2.025231,0.037520,53.977497,0.245061,-5.204969,...,NaN,8.362,8.365,8.322,NaN,NaN,NaN,NaN,NaN,0.083372


In [12]:
Simbad.add_votable_fields('sptype')
Simbad.add_votable_fields('sp_bibcode')
coordinates = SkyCoord(test[["ra","dec"]].values,unit=("deg","deg"))
tab = Simbad.query_region(coordinates, radius=radius*u.arcsec)

ReadTimeout: HTTPConnectionPool(host='simbad.u-strasbg.fr', port=80): Read timed out. (read timeout=60)

In [ ]:
test2=df_2[6000:]
coordinates = SkyCoord(test2[["ra","dec"]].values,unit=("deg","deg"))
tab2 = Simbad.query_region(coordinates, radius=radius*u.arcsec)

In [6]:
Simbad.list_votable_fields()

--NOTES--

1. The parameter filtername must correspond to an existing filter. Filters include: B,V,R,I,J,K.  They are checked by SIMBAD but not astroquery.simbad

2. Fields beginning with rvz display the data as it is in the database. Fields beginning with rv force the display as a radial velocity. Fields beginning with z force the display as a redshift

3. For each measurement catalog, the VOTable contains all fields of the first measurement. When applicable, the first measurement is the mean one. 

Available VOTABLE fields:

bibcodelist(y1-y2)
biblio
cel
cl.g
coo(opt)
coo_bibcode
coo_err_angle
coo_err_maja
coo_err_mina
coo_qual
coo_wavelength
coordinates
dec(opt)
dec_prec
diameter
dim
dim_angle
dim_bibcode
dim_incl
dim_majaxis
dim_minaxis
dim_qual
dim_wavelength
dimensions
distance
distance_result
einstein
fe_h
flux(filtername)
flux_bibcode(filtername)
flux_error(filtername)
flux_name(filtername)
flux_qual(filtername)
flux_system(filtername)
flux_unit(filtername)
fluxdata(filtername)